File: app/src/main/java/com/lazylee/lzywanandroid/tools/Logger.java
package com.lazylee.lzywanandroid.tools;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import com.lazylee.lzywanandroid.BuildConfig;

import java.util.Arrays;
import java.util.List;

/**
 * 可以根据是否在debug模式判断打印log信息
 */
public class Logger {

    private static final String TAG = "Logger";
    private static final boolean DEBUG = BuildConfig.DEBUG;
    private static void e(@NonNull String tag, @Nullable Object text) {
        if (!DEBUG) return;
        Log.e(tag, text != null ? text.toString() : "LOGGER IS NULL");//avoid null
    }

    private static void d(@NonNull String tag, @Nullable Object text) {
        if (!DEBUG) return;
        Log.d(tag, text != null ? text.toString() : "LOGGER IS NULL");//avoid null
    }

    private static void i(@NonNull String tag, @Nullable Object text) {
        if (!DEBUG) return;
        Log.i(tag, text != null ? text.toString() : "LOGGER IS NULL");//avoid null
    }

    public static void d(@Nullable Object text) {
        d(getCurrentClassName() + " || " + getCurrentMethodName(), text);//avoid null
    }

    public static void i(@Nullable Object text) {
        i(getCurrentClassName() + " || " + getCurrentMethodName(), text);//avoid null
    }

    public static void e(Object... objects) {
        if (objects != null && objects.length > 0) {
            e(getCurrentClassName() + " || " + getCurrentMethodName(), Arrays.toString(objects));
        } else {
            e(getCurrentClassName() + " || " + getCurrentMethodName(), getCurrentMethodName());
        }
    }

    public static void e(List<Object> objects) {
        if (objects != null) {
            e(getCurrentClassName() + " || " + getCurrentMethodName(), Arrays.toString(objects.toArray()));
        } else {
            e(TAG, null);
        }
    }

    private static String getCurrentMethodName() {
        try {
            return Thread.currentThread().getStackTrace()[4].getMethodName() + "()";
        } catch (Exception ignored) {}
        return TAG;
    }

    private static String getCurrentClassName() {
        try {
            String className = Thread.currentThread().getStackTrace()[4].getClassName();
            String[] temp = className.split("[.]");
            className = temp[temp.length - 1];
            return className;
        } catch (Exception ignored) {}
        return TAG;
    }
}

File: app/src/main/java/com/lazylee/lzywanandroid/tools/Logger.kt
package com.lazylee.lzywanandroid.tools


import android.util.Log

import com.lazylee.lzywanandroid.BuildConfig

import java.util.Arrays

/**
 * 可以根据是否在debug模式判断打印log信息
 */
object Logger {

    private val TAG = "Logger"
    private val DEBUG = BuildConfig.DEBUG

    private val currentMethodName: String
        get() {
            try {
                return Thread.currentThread().stackTrace[4].methodName + "()"
            } catch (ignored: Exception) {
            }

            return TAG
        }

    private val currentClassName: String
        get() {
            try {
                var className = Thread.currentThread().stackTrace[4].className
                val temp = className.split("[.]".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
                className = temp[temp.size - 1]
                return className
            } catch (ignored: Exception) {
            }

            return TAG
        }

    fun e(tag: String, text: Any) {
        if (!DEBUG) return
        Log.e(tag, text?.toString() ?: "LOGGER IS NULL")//avoid null
    }

    fun d(tag: String, text: Any?) {
        if (!DEBUG) return
        Log.d(tag, text?.toString() ?: "LOGGER IS NULL")//avoid null
    }

    fun i(tag: String, text: Any?) {
        if (!DEBUG) return
        Log.i(tag, text?.toString() ?: "LOGGER IS NULL")//avoid null
    }

    fun d(text: Any?) {
        d("$currentClassName || $currentMethodName", text)//avoid null
    }

    fun i(text: Any?) {
        i("$currentClassName || $currentMethodName", text)//avoid null
    }

    fun e(vararg objects: Any) {
        if (objects != null && objects.size > 0) {
            e("$currentClassName || $currentMethodName", Arrays.toString(objects))
        } else {
            e("$currentClassName || $currentMethodName", currentMethodName)
        }
    }

    fun e(objects: List<Any>?) {
        if (objects != null) {
            e("$currentClassName || $currentMethodName", Arrays.toString(objects.toTypedArray()))
        } else {
            e(TAG)
        }
    }
}
